### DISCUSSION TITLE
Coupling with gradients of Material property

### DISCUSSION THREAD

**Discussion by Ttw0626** [idaholab/moose#27615] [Q&A Modules: Solid mechanics] [answered]
Hi, moose-developers, here is Q about the extension of https://github.com/idaholab/moose/discussions/22910  and https://github.com/idaholab/moose/discussions/24361

as these previous anwsers, I would like to calculate the gradient of material property(scalar) locally which means the additional dofs canbe avoid.

**Q1**
in https://github.com/idaholab/moose/discussions/24361, a new Material object `(MaterialPropertyGradient)` maybe implemented but I could not find it. How can I use this function to get this `MaterialPropertyGradient` like `getMaterialPropertyGradient<RealVectorValue>()` ??

**Q2 if the Q1 not**
I tried the suggestion by gary in his app https://github.com/hugary1995/eel/blob/main/src/materials/chemistry/ChemicalPotential.C#L56-L78, I just changed the derivatives of energys term to local scalar material property. Unfortunately, when I got the gradients of it within `computeQpResidual` or `computeQpJacobian` in kernels (not AD), the nodal result is not continuous. Could you give me
some helps about the gradient of material property in non-AD kernels ?

Thanks
tianwen

**Comment by GiudGiud** [comment]
@hugary1995

**Reply by Ttw0626** [reply]
Hi giud, I  wanna make sure I am not misunderstanding, it will help me to solve this problem maybe.

**Q1**
If I wanna do some operations locally in one element, is it correct focus on `computeProperties` ?
after the Qps loop on `computeQpProperty()`.
so that, I can get material properties (scalar) on all of Qps in this element using `oneMaterialProperty[_qp]`.

**Q2: Id of Qps in one element**
is it correct using just `_qp` ?

**Q3: current coordinates of one Qp in one element**
is it correct using just '_q_point[_qp] (x/y/z)' ?

Thanks.
tianwen

**Reply by GiudGiud** [reply, marked as answer]
Q1 for most applications you are meant to implement computeQpProperties not computeProperties

If you want to override the computeProperties, yes you will have access to the property at every Qp after a loop calling computeQpProperties

**Reply by GiudGiud** [reply]
Q2 
Yes the index in an element of the properties is _qp

**Reply by GiudGiud** [reply]
Q3 that s correct though x should index 0, y index 1 etc

**Reply by Ttw0626** [reply]
guid, Thanks for these, I understood. That's make sense to me.

tianwen
